src/main/java/org/testng/JarFileUtils.java

```diff
@@ -52,23 +57,40 @@ private boolean testngXmlExistsInJar(File jarFile, List<String> classes) throws
         boolean foundTestngXml = false;
         try (JarFile jf = new JarFile(jarFile)) {
             Enumeration<JarEntry> entries = jf.entries();
+            File file = java.nio.file.Files.createTempDirectory("testngXmlPathInJar-").toFile();
+            String suitePath = null;
             while (entries.hasMoreElements()) {
                 JarEntry je = entries.nextElement();
-                if (matchesXmlPathInJar(je)) {
-                    Collection<XmlSuite> parsedSuites = Parser.parse(jf.getInputStream(je), processor);
-                    for (XmlSuite suite : parsedSuites) {
-                        // If test names were specified, only run these test names
-                        XmlSuite suiteToAdd = suite;
-                        if (testNames != null) {
-                            suiteToAdd = XmlSuiteUtils.cloneIfContainsTestsWithNamesMatchingAny(suite, testNames);
-                        }
-                        suites.add(suiteToAdd);
+                String jeName = je.getName();
+                if (Parser.canParse(jeName.toLowerCase())){ 
+                    InputStream  inputStream = jf.getInputStream(je);
+                    File copyFile = new File(file, jeName);
+                    Files.copyFile(inputStream, copyFile);
+                    if (matchesXmlPathInJar(je)) {
+                        suitePath = copyFile.toString();
                     }
-                    return true;
                 } else if (isJavaClass(je)) {
                     classes.add(constructClassName(je));
                 }
             }
+            if (Strings.isNullOrEmpty(suitePath)) {
+                return foundTestngXml;
+            }
+            Collection<XmlSuite> parsedSuites = Parser.parse(suitePath,processor); 
+            for (XmlSuite suite : parsedSuites) {
+                // If test names were specified, only run these test names
+                if (testNames != null) {
+                  TestNamesMatcher testNamesMatcher = new TestNamesMatcher(suite, testNames);
+                  List<String> missMatchedTestname = testNamesMatcher.getMissMatchedTestNames();
+                  if (!missMatchedTestname.isEmpty()) {
+                    throw new TestNGException("The test(s) <" + Arrays.toString(missMatchedTestname.toArray())+ "> cannot be found.");
```

Comment:
Can we not just do `throw new TestNGException("The test(s) <" + missMatchedTestname + "> cannot be found.");` ? Why do the `Arrays.toString()` and within which we are again doing a `List.toArray()`. It looks redundant.

Reply:
I have the same question but I suppose the string output is not exactly the same

Reply:
Done